### DISCUSSION TITLE
Patroni to use CA Certs for replication user

### DISCUSSION THREAD

**Discussion by djNutanix** [patroni/patroni#3193] [Q&A] [answered]
Hi

We configured CA cert authentication [password less] for replication user and manually verified that its working using psql. But seems like Patroni has mandatory username and password fields under the "authentication" section for replication and even if we provide the sslmode, sslcert, sslkey & sslrootcert parameters along with username & password, it's connecting using the username/password only and not with the certs. If I remove the password from the patroni.yml then Patroni complains about it:

```
2024-10-22 16:28:47,446 ERROR: Can not fetch local timeline and lsn from replication connection
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/patroni/postgresql/__init__.py", line 1078, in get_replica_timeline
    with self.get_replication_connection_cursor(**self.config.local_replication_address) as cur:
  File "/usr/lib64/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/usr/local/lib/python3.6/site-packages/patroni/postgresql/__init__.py", line 1073, in get_replication_connection_cursor
    with get_connection_cursor(**conn_kwargs) as cur:
  File "/usr/lib64/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/usr/local/lib/python3.6/site-packages/patroni/postgresql/connection.py", line 157, in get_connection_cursor
    conn = psycopg.connect(**kwargs)
  File "/usr/local/lib/python3.6/site-packages/patroni/psycopg.py", line 103, in connect
    ret = _connect(*args, **kwargs)
  File "/usr/local/lib64/python3.6/site-packages/psycopg2/__init__.py", line 127, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: fe_sendauth: no password supplied
```

It will be really helpful if you can explain how to achieve password less authentication using CA Certs for replication user with Patroni.

Thanks
Daljit Singh

**Comment by CyberDem0n** [comment, marked as answer]
There is no requirement to have passwords in authentication section. In case if password is missing Patroni is just trying connecting without using it.

When looking at the error I could only guess that you didn't properly configured pg_hba.conf to allow such kinds of connections.
Also, with "local" replication connections there could be some problems, because Patroni is usually trying to use "localhost" as a hostname for them, and usually nobody adds the "localhost" to alternatives in certificate, because it is bad security wise. The last one is recommended to solve by using `postgresql.use_unix_socket: on` and `postgresql.use_unix_socket_repl: on` parameters, so that Patroni will connect to the local postgres using unix sockets.

**Comment by djNutanix** [comment]
Thank you very much for pointing me to the right direction. The error to me was bit confusing because it was focusing on "no password" and even the Patroni document doesn't explicitly mention that the password can be optional if Certs are used. 

After fixing the pg_hba.conf, I was able to get Patroni working with Certs without any password.
